scripts/gen_mmu.py

```diff
@@ -33,36 +40,941 @@
+    # This returns the size needed for the given set of page tables.
+    def set_binary_file_size(self):
+        binary_size = ctypes.create_string_buffer((4096)+
+                                                  (len(list_of_pde.keys()) *
+                                                   4096))
+        return( binary_size)
+
+    #prints the details of the pde
+    def verbose_output(self):
+
+        print("\nTotal Page directory entries " + str(self.get_number_of_pd()))
+        count =0
+        for key, value in list_of_pde.items():
+            for i in value.page_entries_info:
+                count+=1
+                print("In Page directory entry "+format_string(value.pde_index) +\
+                      ": valid start address = "+ \
+                      hex_32(i.start_addr) + ", end address = " + \
+                      hex_32((i.pte_valid_addr_end +1 )*4096 -1 +\
+                             (value.pde_index * (FourMB))))
+
+    # print all the tables for a given page table mode
+    def print_all_page_table_info(self):
+        self.pde_print_elements()
+        self.pte_print_elements()
+
+    def pde_verbose_output(self,pde, binary_value):
+        if args.verbose == False:
+            return
+
+        global print_string_pde_list
+
+        present = format_string(binary_value & 0x1 )
+        read_write = format_string((binary_value >> 1 ) & 0x1 )
+        user_mode = format_string((binary_value >> 2 ) & 0x1 )
+        pwt = format_string((binary_value >> 3 ) & 0x1 )
+        pcd = format_string((binary_value >> 4 ) & 0x1 )
+        a = format_string((binary_value >> 5 ) & 0x1 )
+        ignored1 = format_string(0)
+        ps = format_string((binary_value >> 7 ) & 0x1 )
+        ignored2 = format_string(0000)
+        page_table_addr = format_string( hex((binary_value >> 12 ) & 0xFFFFF))
+
+        print_string_pde_list += ( format_string(str(pde))+" | "+(present)+ " | "+\
+              (read_write)+ " | "+\
+              (user_mode)+ " | "+\
+              (pwt)+ " | "+\
+              (pcd)+ " | "+\
+              (a)+ " | "+\
+              (ps)+ " | "+
+              page_table_addr +"\n"
+        )
+
+    def pde_print_elements(self):
+        global print_string_pde_list
+        print("PAGE DIRECTORY ")
+        print(format_string("PDE")+" | "+ \
+              format_string('P')  +" | "+  \
+              format_string('rw')   +" | "+  \
+              format_string('us')  +" | "+  \
+              format_string('pwt')  +" | "+  \
+              format_string('pcd')  +" | "+  \
+              format_string('a')  +" | "+   \
+              format_string('ps')  +" | "+  \
+              format_string('Addr page table'))
+        print(print_string_pde_list)
+        print("END OF PAGE DIRECTORY")
+
+
+    def pte_verbose_output(self, pde, pte, binary_value):
+        global pde_pte_string
+
+        present    = format_string( str((binary_value >> 0) & 0x1))
```

Comment:
Curious why sometimes `str()` is called when passing an argument to `format_string()`, and sometimes it's not, even though numeric arguments are passed in both cases (for instance, in `pde_verbose_output()`).